userguide/storage.rst

```diff
@@ -713,53 +713,70 @@ existing datasets to close them--the remaining
 Deduplication
 ^^^^^^^^^^^^^
 
-Deduplication is the process of ZFS transparently reusing a single
-copy of duplicated data to save space. Depending on the amount of
-duplicate data, deduplicaton can improve storage capacity, as less
-data is written and stored. However, deduplication is RAM intensive. A
-general rule of thumb is 5 GB of RAM per terabyte of deduplicated
-storage. **In most cases, compression provides storage gains
-comparable to deduplication with less impact on performance.**
-
-In %brand%, deduplication can be enabled during dataset creation. Be
-forewarned that **there is no way to undedup the data within a dataset
-once deduplication is enabled**, as disabling deduplication has
-**NO EFFECT** on existing data. The more data written to a
-deduplicated dataset, the more RAM it requires. When the system starts
-storing the DDTs (dedup tables) on disk because they no longer fit
-into RAM, performance craters. Further, importing an unclean pool can
-require between 3-5 GB of RAM per terabyte of deduped data, and if the
-system does not have the needed RAM, it will panic. The only solution
-is to add more RAM or recreate the pool.
-**Think carefully before enabling dedup!**
-This `article
-<http://constantin.glez.de/blog/2011/07/zfs-dedupe-or-not-dedupe>`_
-provides a good description of the value versus cost considerations
-for deduplication.
-
-**Unless a lot of RAM and a lot of duplicate data is available, do not
-change the default deduplication setting of "Off".**
-For performance reasons, consider using compression rather than
-turning this option on.
-
-If deduplication is changed to *On*, duplicate data blocks are removed
-synchronously. The result is that only unique data is stored and
-common components are shared among files. If deduplication is changed
-to *Verify*, ZFS will do a byte-to-byte comparison when two blocks
-have the same signature to make sure that the block contents are
-identical. Since hash collisions are extremely rare, *Verify* is
-usually not worth the performance hit.
-
-.. note:: After deduplication is enabled, the only way to disable it
-   is to use the :samp:`zfs set dedup=off {dataset_name}` command
-   from :ref:`Shell`. However, any data that has already been
-   deduplicated will not be un-deduplicated. Only newly stored data
-   after the property change will not be deduplicated. The only way to
-   remove existing deduplicated data is to copy all of the data off of
-   the dataset, set the property to off, then copy the data back in
-   again. Alternately, create a new dataset with
-   :guilabel:`ZFS Deduplication` left disabled, copy the data to the
-   new dataset, and destroy the original dataset.
+#ifdef freenas
+.. index:: Deduplication
+.. _Deduplication:
+
+Deduplication
+^^^^^^^^^^^^^
+
+Deduplication ("dedup") is a process where ZFS aims to store a single copy
+of any data that is repeated in different blocks or files, rather than storing 
```

Comment:
Condense this down to "a single copy of block data" From what I know, ZFS only does deduplication of block data, not file deduplication at all.

Reply:
For similar reason I'm not happy about the requested change to refer only to "block data". Yes technically ZFS dedups on blocks. No, end-users will not necessarily know what that means, nor should they be expected to if not using a zvol, why confuse them? If the user is storing files, then it's deduping files, albeit by means of deduping the blocks that constitute the files. *(A bit like saying the OS doesn't read a file, it's reading sectors that make up a file. Technically true, in practice pointless for "guide" purposes)*

Reply:
The problem is that file-level dedup is almost always what leads users to think they need it.  And then they regret it.  I'm in favor of simplifying things, but not to the point that it hides a crucial fact.  And the blocks used by two files are not guaranteed to be identical.  One could be offset due to block sub-allocation or something.

Reply:
In that case the issue is the warnings, not the belief in file-level dedup vs block level dedup. We aren't warning of the right things - which is that file level dedup may not realise expected benefits since it doesan't work that way. We don't need to go into how it actually does work, to make that point with more force.

We might instead warn that *dedup does not always deduplicate even apparently-identical files, as it works at the level of blocks rather than files.  **It is imperative to actually check the benefits of deduplication and its impact on performance, even with highly duplicated data. Do not assume any benefits might be realised from enabling this setting, unless tests have actually confirmed this to be the case**.*

Would something like that work? The whole issue of block vs file is a red herring it seems - the user doesn't need to know exactly how it works. They need to know that what they might assume, isn't so, and not to assume it. A brief mention of blocks (as above) is sufficient to underpin that point.

Reply:
In that case the issue is the warnings. It's not the lack of education about file-level dedup vs block level dedup. The problem is, we're not warning of the right things, or not warning of them adequately - namely that file level dedup may not realise the naively expected benefits since ZFS doesn't actually work that way. But we don't need to go in depth about how ZFS actually does work, to make that point with necessary force.

We might instead warn that: *The deduplication function does not always deduplicate even apparently-identical files, because it works at the level of data blocks on disk, and not on individual files.  **Therefore it is imperative to actually check the benefits of deduplication and its impact on performance, even with highly duplicated data, before assuming that deduplication will provide any benefits whatsoever. The number of cases where deduplication is actually useful, is small. Do not assume any benefits will be realised from enabling this setting, in any situation, unless your own tests have actually confirmed this to be the case**.*

Would something like that work? The whole issue of block vs file is a red herring it seems - the user doesn't need to know exactly how it works. They need to know that what they might assume, isn't so, and not to assume it. A brief mention of blocks (as above) is sufficient to underpin that point.